arangod/Aql/OptimizerRulesReplaceFunctions.cpp

```diff
@@ -0,0 +1,464 @@
+struct fulltextParams{
+  std::string collection;
+  std::string attribute;
+  AstNode* limit = nullptr;
+
+  fulltextParams(AstNode const* node){
+    TRI_ASSERT(node->type == AstNodeType::NODE_TYPE_FCALL);
+    AstNode* arr = node->getMember(0);
+    TRI_ASSERT(arr->type == AstNodeType::NODE_TYPE_ARRAY);
+    if (arr->getMember(0)->isStringValue()){
+      collection = arr->getMember(0)->getString();
+    }
+    if (arr->getMember(0)->isStringValue()){
+      attribute = arr->getMember(1)->getString();
+    }
+    if(arr->numMembers() > 3){
+      limit = arr->getMember(3);
+    }
+  }
+};
+
+AstNode* getAstNode(CalculationNode* c){
+  return c->expression()->nodeForModification();
+}
+
+Function* getFunction(AstNode const* ast){
+  if (ast->type == AstNodeType::NODE_TYPE_FCALL){
+    return static_cast<Function*>(ast->getData());
+  }
+  return nullptr;
+}
+
+AstNode* createSubqueryWithLimit(
+  ExecutionPlan* plan,
+  ExecutionNode* node,
+  ExecutionNode* first,
+  ExecutionNode* last,
+  Variable* lastOutVariable,
+  AstNode* limit
+  ){
+  // Creates a subquery of the following form:
+  //
+  //    singleton
+  //        |
+  //      first
+  //        |
+  //       ...
+  //        |
+  //       last
+  //        |
+  //     [limit]
+  //        |
+  //      return
+  //
+  // The subquery is then injected into the plan before the given `node`
+  // This function returns an `AstNode*` of type reference to the subquery's
+  // `outVariable` that can be used to replace the expression (or only a
+  // part) of a `CalculationNode`.
+  //
+  auto* ast = plan->getAst();
+
+  /// singleton
+  ExecutionNode* eSingleton = plan->registerNode(
+      new SingletonNode(plan, plan->nextId())
+  );
+
+  /// return
+  ExecutionNode* eReturn = plan->registerNode(
+      // link output of index with the return node
+      new ReturnNode(plan, plan->nextId(), lastOutVariable)
+  );
+
+  /// link nodes together
+  first->addDependency(eSingleton);
+  eReturn->addDependency(last);
+
+  /// add optional limit node
+  if(limit && !limit->isNullValue()) {
+    ExecutionNode* eLimit = plan->registerNode(
+      new LimitNode(plan, plan->nextId(), 0 /*offset*/, limit->getIntValue())
+    );
+    plan->insertAfter(last, eLimit); // inject into plan
+  }
+
+  /// create subquery
+  Variable* subqueryOutVariable = ast->variables()->createTemporaryVariable();
+  ExecutionNode* eSubquery = plan->registerSubquery(
+      new SubqueryNode(plan, plan->nextId(), eReturn, subqueryOutVariable)
+  );
+
+  plan->insertBefore(node, eSubquery);
+
+  // return reference to outVariable
+  return ast->createNodeReference(subqueryOutVariable);
+
+}
+
+AstNode* replaceNearOrWithin(AstNode* funAstNode, ExecutionNode* calcNode, ExecutionPlan* plan, bool isNear){
+  auto* ast = plan->getAst();
+  auto* query = ast->query();
+  auto* trx = query->trx();
+  nearOrWithinParams params(funAstNode,isNear);
+  bool supportLegacy = true;
+  (void) supportLegacy;
```

Comment:
I'm not sure I understand what is happening here with `supportLegacy`. Does this just need to be cleaned up?